python/graphframes/examples.py

```diff
@@ -57,3 +60,216 @@ def friends(self):
+    Thin Junction Trees." AISTATS, 2011.
+
+    The BP algorithm works by:
+
+    * Coloring the graph by assigning a color to each vertex such that no neighboring vertices
+      share the same color.
+    * In each step of BP, update all vertices of a single color.  Alternate colors.
+     """
+    @classmethod
+    def main(cls):
+        """Run the algorithm for an example problem."""
+        conf = SparkConf().setAppName("BeliefPropagation example")
+        sc = SparkContext.getOrCreate(conf)
+        sql = SQLContext.getOrCreate(sc)
+
+        # create graphical model g of size 3 x 3
+        g = Graphs(sql).gridIsingModel(3)
+        print("Original Ising model:")
+        g.vertices.show()
+        g.edges.show()
+
+        # run BP for 5 iterations
+        numIter = 5
+        results = cls.runBPwithGraphFrames(g, numIter)
+
+        # display beliefs
+        beliefs = results.vertices.select('id', 'belief')
+        print("Done with BP. Final beliefs after {} iterations:".format(numIter))
+        beliefs.show()
+
+        sc.stop()
+
+    @staticmethod
+    def colorGraph(g):
+        """Given a GraphFrame, choose colors for each vertex.
+
+        No neighboring vertices will share the same color. The number of colors is minimized.
+
+        This is written specifically for grid graphs. For non-grid graphs, it should be generalized,
+        such as by using a greedy coloring scheme.
+
+        :param g: Grid graph generated by :meth:`Graphs.gridIsingModel()`
+        :return: Same graph, but with a new vertex column "color" of type Int (0 or 1)
+
+        """
+
+        colorUDF = sqlFuncs.udf(
+            lambda i, j: 0 if (i + j) % 2 == 0 else 1,
+            returnType=types.IntegerType())
+        v = g.vertices.withColumn('color', colorUDF(sqlFuncs.col('i'), sqlFuncs.col('j')))
+        return GraphFrame(v, g.edges)
+
+    @classmethod
+    def runBPwithGraphFrames(cls, g, numIter):
+        """Run Belief Propagation using GraphFrame.
+
+        This implementation of BP shows how to use GraphFrame's aggregateMessages method.
+        """
+        # choose colors for vertices for BP scheduling
+        colorG = cls.colorGraph(g)
+        numColors = colorG.vertices.select('color').distinct().count()
+
+        # TODO: handle vertices without any edges
+
+        # initialize vertex beliefs at 0.0
+        gx = GraphFrame(colorG.vertices.withColumn('belief', sqlFuncs.lit(0.0)), colorG.edges)
+
+        # run BP for numIter iterations
+        for iter_ in range(numIter):
+            # for each color, have that color receive messages from neighbors
+            for color in range(numColors):
+                # Send messages to vertices of the current color.
+                # We may send to source or destination since edges are treated as undirected.
+                msgForSrc = sqlFuncs.when(
+                    AM.src()['color'] == color,
+                    AM.edge()['b'] * AM.dst()['belief'])
```

Comment:
I vote for 2 or 3, both look good to me.

Maybe I took the conversation off-track by using the word "singleton", sorry. We don't have the issues normally associated with singletons in the PR because there is no reason for us to try and prevent the user from initializing their own version of `_AggregateMessages`. If some user wants to do that for some reason, it's perfectly fine. However, I think we should provide an initialized version for most users who don't need to know the implementation details, they simply need a name space that'll allow them to use `AM.src` and `AM.dst`. I think both option 2 and option 3 do that quite well.